pkl-cli/pkl-cli.gradle.kts

```diff
@@ -1,28 +1,107 @@
+// https://youtrack.jetbrains.com/issue/KTIJ-19369
+@file:Suppress("DSL_SCOPE_VIOLATION", "UnstableApiUsage")
+
+import org.gradle.configurationcache.extensions.capitalized
+
+
 plugins {
-  pklAllProjects
-  pklKotlinLibrary
-  pklPublishLibrary
-  pklNativeBuild
+  id("pklAllProjects")
+  id("pklJvmEntrypoint")
+  id("pklPureKotlin")
+  id("pklPublishLibrary")
+  id("pklNativeBuild")
   `maven-publish`
+  distribution
 
-  // already on build script class path (see buildSrc/build.gradle.kts),
-  // hence must only specify plugin ID here
-  @Suppress("DSL_SCOPE_VIOLATION")
   id(libs.plugins.shadow.get().pluginId)
-
-  @Suppress("DSL_SCOPE_VIOLATION")
   alias(libs.plugins.checksum)
 }
 
+description = "Pkl command line interface entrypoint"
+private val entrypoint = "org.pkl.cli.Main"
+private val module = "pkl.cli"
+
 // make Java executable available to other subprojects
 val javaExecutableConfiguration: Configuration = configurations.create("javaExecutable")
 
+// make Java executable available to other subprojects
+val nativeCliExecutableConfiguration: Configuration = configurations.create("nativeExecutable")
+
+val enablePgo = false
+val oracleGvm = false
+val pgoInstrument = enablePgo
+val enableExperimental = true
+val profileName = "pkl-cli.iprof"
+val profilesZip = layout.projectDirectory.file("pgo.zip")
+val enableRelease = findProperty("nativeRelease") == "true"
+val isNativeBuildEnabled = gradle.startParameter.taskNames.any { subject ->
+  subject.lowercase().let {
+    "native" in it || "executable" in it
+  }
+}
+val extraJavacArgs: List<String> = listOfNotNull(
+  "--add-exports=org.graalvm.truffle.runtime.svm/com.oracle.svm.truffle=$module",
+  if (oracleGvm) null else "--add-reads=org.graalvm.truffle.runtime.svm=$module",
+  if (!oracleGvm) null else "--add-reads=com.oracle.svm.truffle=$module",
+  "--add-reads=$module=ALL-UNNAMED",
+)
+val extraJvmArgs: List<String> = listOf(
+  "--enable-native-access=ALL-UNNAMED",
+)
+
+val devNativeImageFlags = listOfNotNull(
+  "-Ob",
+  if (!enablePgo) null else "--pgo-instrument",
+)
```

Comment:
Native Image now builds by default with `-Ob` (optimizes build time). To build a "release" native binary, pass `-PnativeRelease=true`.

When built in dev mode, the binary is instrumented with `--pgo-instrument` when `enablePgo` (a flag in this script) is flipped to `true`.